Clarify and organize cross browser testing strategies

kevinlu11 opened the pull request:
Made them more concise and focused, especially the introduction. Improved phrasing.

<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

kevinlu11 committed:
Clarify and organize first few sections

Made them more concise and focused, especially the introduction. Improved phrasing.

wbamberg commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
+### Determine the most commonly-used browsers among your target audience
 
-You could call these "assumptions", or "gut feelings". This is not an accurate, scientific approach, but as someone who has experience with the web industry you'll have a pretty good idea of at least some of the browsers you should test. This can form a good basis for a support chart.
+Most people in North America or Western America use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.
```
```suggestion
Most people in North America or Western Europe use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.
```

Josh-Cena replied:
Edge is practically the same as Chrome, so it should be fine if you don't test it at all.

klu22 replied:
Agreed on all points; I just lumped Edge with IE to preserve the original grouping, and think it would be a good idea to change it, now that you mention it, given how much more modern Edge is than IE.

wbamberg commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
+Opera Mini isn't very capable in terms of running complex JavaScript at runtime. See [Opera Mini and JavaScript](https://dev.opera.com/articles/opera-mini-and-javascript/) for more details. We should put this into grade B as well.
 
-You might also know that a number of people still use IE 9. This is old and less capable, so let's put it in the B grade tier.
 
```
```suggestion
```

(fix for markdownlint)

wbamberg reviewed (changes requested):
Thank you @klu22 ! Quite often I'm sceptical about these copy-editing PRs, but this one looks like a great improvement, I agree with pretty much all your suggested changes 👍 .

Josh-Cena commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
+## Choosing which browsers and devices to test
 
-When doing cross-browser testing, you need to work out a list of browsers you will need to test on to start with. There is no way you can test on every combination of browser and device your users might use to view your site — there are just too many, and new ones appear all the time.
+Since it's impossible to test every combination of browser and device, it's enough that you ensure your site works on the most important ones. In practical applications, "important" often means "commonly used among the target audience."
```
```suggestion
Since it's impossible to test every combination of browser and device, it's enough that you ensure your site works on the most important ones. In practical applications, "important" often means "commonly used among the target audience".
```

klu22 replied:
This is a regional difference--periods go inside quotes in American English, but outside quotes in British English. I don't know whether MDN is read by more British English users or American English users.

Josh-Cena replied:
Periods don't go into the quotes if the quote is not a full sentence. I'm okay with (though I also won't write):

> He said, "I love you."

Josh-Cena replied:
According to a whacky regex test over the corpus of MDN, we seem to use `".` much more commonly than `."` (>2000 times to >500 times). Still, it may be because more than half of my life's academic English training had been in Oxford English and it's just stuck with me subconsciously—I'm fine if you want to keep it this way. We can have an editorial consensus later since we have a lot of stylistic conventions to settle.

wbamberg replied:
Overall on MDN we're not very prescriptive on fine style guide points like this. I would say in general, that if the [writing style guide](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Guidelines/Writing_style_guide) isn’t explicit on a point, and if it’s not really obviously right or wrong one way or the other (to the point where it affects comprehension, I mean), then it doesn’t matter much, and it’s not really worthwhile to change docs from one to the other.

For our docs, clarity, helpfulness, and correctness are much more important than fine style guide points.

Josh-Cena commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
 
 1. A grade: Common/modern browsers — Known to be capable. Test thoroughly and provide full support.
 2. B grade: Older/less capable browsers — known not to be capable. Test, and provide a more basic experience that gives full access to core information and services.
 3. C grade: Rare/unknown browsers — don't test, but assume they are capable. Serve the full site, which should work, at least with the fallbacks provided by our defensive coding.
```
You deleted the paragraph about defensive coding so this seems out of context. I suggest restoring that paragraph about graceful degradation.

klu22 replied:
Defensive coding is irrelevant to this paragraph, which is about choosing which browsers to test.

Josh-Cena commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
+### Determine the most commonly-used browsers among your target audience
 
-You could call these "assumptions", or "gut feelings". This is not an accurate, scientific approach, but as someone who has experience with the web industry you'll have a pretty good idea of at least some of the browsers you should test. This can form a good basis for a support chart.
+Most people in North America or Western America use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.
```
This is far too America-centric for an opening paragraph—it's only supposed to be an example. If you find the original opening too informal, I suggest rewording it but at least keeping the same idea that we are "making educated guesses".

klu22 replied:
The bulk of MDN readers are based in either North America or Western Europe, and those who aren't are free to find information about their respective region. As you say, it's only supposed to be an example, not a comprehensive list of most-used browsers by region. 

I disagree that "making educated guesses" should be kept. Making educated guesses is merely a means to an end (the end being determining which browsers and devices one should test), and telling someone "make educated guesses" doesn't tell them anything new.

klu22 replied:
On second thought: an extra opening sentence seems redundant, given the section heading.

I'll preface the first sentence with "For example" and leave it at that. Flow is extremely important to me.

klu22 commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,54 @@ In this article we drill down into carrying out testing, looking at identifying
 ### Browser support stats
 
-One helpful measure you can call on to inform your browser testing choices is browser support stats. There are a number of sites that provide such stats, for example:
+Browser support stats can make your . For example:
```
Good point, will amend

kevinlu11 committed:
Fix typos, incorporate suggestions from others

Added "For example" at the beginning of the section "Determine the most commonly-used browsers[...]"
Simplified some wording even further

kevinlu11 reviewed (commented):
I agree with the proposed changes, except where noted in the relevant conversations under the edit suggestions. Thanks for the feedback guys.

kevinlu11 committed:
Merge branch 'main' into patch-4

wbamberg reviewed (changes requested):
I do want to have the preceding sentence here, to emphasise that this is audience-relative:

> Deciding which browsers to test is a matter of educated guesses, and should be based on understanding of your user's demographics. 

Otherwise I'm 👍 with these changes.

kevinlu11 committed:
Merge branch 'main' into patch-4

kevinlu11 committed:
Improve style

Renamed the section 
"Determine the most commonly-used browsers among your target audience" 
to 
"Determine your audience's most commonly used browsers"

As per others' suggestions, added an intro sentence to this section for extra context

kevinlu11 committed:
Change "Determine" to "Predict" in the second section title

"Predict" seems more consistent with the theme of educated guesses.

Reworded some things, added further context

kevinlu11 reviewed (commented):
I've added an intro sentence along those lines; I also changed the section title from "Determine [...]" to "Predict [...]" as I feel the latter fits the theme of educated guesses more.

kevinlu11 committed:
Merge branch 'main' into patch-4

kevinlu11 committed:
Remove trailing whitespace

kevinlu11 committed:
Merge branch 'main' into patch-4

Josh-Cena commented on files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md:
```diff
@@ -44,63 +44,52 @@ In this article we drill down into carrying out testing, looking at identifying
-By coding defensively, we mean trying to build in intelligent fallbacks so that if a feature or style doesn't work in a browser, the site will be able to downgrade to something less exciting that still provides an acceptable user experience — the core information is still accessible, for example, even if it doesn't look quite as nice.
-
-The aim is to build up a chart of browsers/devices you can refer to as you test. You can make this as simple or as complex as you like — for example a common approach is to have multiple grades of support level, something like:
+You can classify browsers and devices by the amount of support you intend to give, e.g.:
```
```suggestion
You can classify browsers and devices by the amount of support you intend to give. For example:
```

Josh-Cena reviewed (approved):
Editorial-wise, looks good to me. @wbamberg you may merge if the content is fine to you as well.

Josh-Cena committed:
Update files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md

Josh-Cena committed:
Merge branch 'main' into patch-4

wbamberg committed:
Merge branch 'main' into patch-4
